### DISCUSSION TITLE
How to scope the atom?

### DISCUSSION THREAD

**Discussion by devnomic** [Omnistac/zedux#48] [General]
I just stumbled upon this lib, coming from jotai, how do i scope the atom like `jotai-molecules` does? https://github.com/saasquatch/jotai-molecules

e.g. you have list component,
then in each list item, it accept initial value and store it in it's own atom, and derive another atom from there.
````
<List>
   <ListItem initialValue={'a'} />
   <ListItem initialValue={'b'} />
</List>
````

**Comment by devnomic** [comment]
After reading the docs it seems using ecosystem is the answer?

**Reply by bowheart** [reply]
Hey, welcome! That is a great question.

The concepts are not exactly the same in Zedux, but you can get similar functionality using one of two different base approaches - using ecosystems or using atoms.

## Using Ecosystems

Yes, ecosystems are what we call the "molecular" part of Zedux. However, ecosystems and [`<EcosystemProvider />`](https://omnistac.github.io/zedux/docs/api/components/EcosystemProvider) are most commonly used like Recoil's `<RecoilRoot>` or React Query's `<QueryClientProvider>` - you typically create only one at the root of your app. You don't have to do it that way, in fact one of the potential features we've considered adding is "state versioning", which encourages branching and merging multiple ecosystems.

An early prototype of Zedux had a Molecule primitive and we got rid of it 'cause Ecosystems can handle everything molecules could do. Think of ecosystems more like a group of molecules than a single molecule. You can define your own rules for what makes a molecule.

For example, you can use an id prefix like `molecule-a/` in several atom keys

```ts
const exampleAtom = atom('molecule-a/example', ...)
```

(**This is just for the sake of example and may not be the best way to create "molecules" in an ecosystem**)

Then use [`ecosystem.findAll('molecule-a')`](https://omnistac.github.io/zedux/docs/api/classes/Ecosystem#findall) to get a list of all instances of atoms in "molecule-a". You can also use atom factories or injectors to create/reuse atom templates with a given prefix based off the prefix for the current atom, but I don't want to get too far into that, because I'd currently recommend the other approach:

## Using Atoms

I've never used the ecosystem approach. All I've ever needed I've been able to do with atoms. The relevant Zedux concepts are [atom params](https://omnistac.github.io/zedux/docs/walkthrough/atom-instances#atom-params), [`<AtomProvider>`](https://omnistac.github.io/zedux/docs/api/components/AtomProvider), and [`useAtomContext()`](https://omnistac.github.io/zedux/docs/api/hooks/useAtomContext). These are the Zedux features that are most like Jotai Molecules right now.

In other words, you should be able to use Zedux atoms to accomplish everything Jotai molecules can do. An atom can _be_ a "molecule". Zedux atoms don't have a concept of Scopes, but you can use atom params and/or an exported ref to configure any data you want for an atom. An atom's state can be a list of other atom instances or atom templates. Atoms can also take other atoms as params.

These can be used in several ways to accomplish similar things to Jotai Molecules. Here's one way to do it using just atom params:

```ts
// a custom injector for creating atom templates for a molecule
const injectMoleculeAtom = (key, value) => {
  // assume first param has a `molecule` property:
  const {
    params: [{ molecule }],
  } = injectSelf()

  return injectMemo(() => atom(`${molecule}/${key}`, value), [])
}

const moleculeScopeAtom = atom('moleculeScope', ({ molecule }) => {
  // create a counterAtom template scoped to this molecule:
  const counterAtom = injectMoleculeAtom('counter', 0)

  return {
    counterAtom,
    scope: { molecule },
  }
})

function ListItem({ initialValue }) {
  const { counterAtom } = useAtomValue(moleculeScopeAtom, [
    { molecule: initialValue },
  ])
  // use the current molecule's counterAtom:
  const [count, setCount] = useAtomState(counterAtom)

  return (
    <div>
      Count: {count}{' '}
      <button onClick={() => setCount(state => state + 1)}>Increment</button>
    </div>
  )
}
```

This can be expanded to use

```ts
useAtomInstance(moleculeScopeAtom, [{ molecule: initialValue }])
```

and `<AtomProvider>` to provide the `moleculeScope` atom instance to all the `<ListItem>`'s children.

If the `initialValue` prop is not necessarily a unique id for the molecule, you can use React's `useId()` hook inside `<ListItem>`.

Lastly, I just wanted to reiterate that Zedux is just a different tool. The above example mimics Jotai Molecules, but it isn't necessarily the cleanest or easiest way to accomplish this sort of thing using Zedux. The theoretical best way to do it depends on the situation. Most situations likely don't need anything beyond a simple `<AtomProvider>`. Or some things might need even more power e.g. using exported refs to attach stuff to the "molecule" atom from a component or inside an effect.

**Reply by bowheart** [reply]
Here's a codesandbox demonstrating the atom-as-molecule approach: https://codesandbox.io/s/zedux-atoms-as-jotai-molecules-3ilmtt

Let me know if anything isn't clear. Best of luck!

**Reply by devnomic** [reply]
Thank you for your very detailed answer! So zedux use atom params to create unique atom, very cool.

I tried to create it the zedux way, is this typically how to write it in zedux?
https://codesandbox.io/s/great-snyder-g4k64w?file=/src/App.tsx

Does zedux also have `useSetAtom` kinda like shortcut to update atom without rerender? without using `useAtomInstance`

**Reply by bowheart** [reply]
That looks great, good use of exports

> Does zedux also have `useSetAtom`

Not currently. We have some 30-odd helpers like that that we use at Omnistac that aren't in Zedux proper. We may add some of the more commonly-requested ones to Zedux, so I'll add a vote to that tally if you want.

But yes, `useAtomInstance` is the way to do it currently. It's very versatile at the cost of being maybe a little low-level. If you want, you can make your own higher-level hooks on top of it for now. A fancy, overloaded `useSetAtom` might look like this:

```ts
export const useSetAtom: {
  <A extends AnyAtomTemplate>(
    template: A,
    params: AtomParamsType<A>,
    config?: ZeduxHookConfig
  ): AtomInstanceType<A>['setState']

  <A extends AnyAtomTemplate>(
    template: ParamlessTemplate<A>
  ): AtomInstanceType<A>['setState']
} = <A extends AnyAtomTemplate>(
  template: A,
  params?: AtomParamsType<A>,
  config: ZeduxHookConfig = {
    operation: OPERATION,
  }
) => useAtomInstance(template, params as AtomParamsType<A>, config).setState

// in a component:
const setMyParamlessAtom = useSetAtom(myParamlessAtomTemplate)
const setMyAtom = useSetAtom(myAtomTemplate, ['with', 'params'])
```

As for your question in the sandbox:

> // Is writing as initialValue as params twice unavoidable?

It can be avoided. Every API in Zedux that takes an AtomTemplate + params combo can also be given an atom instance.

Before:

```ts
const { decrement, increment } = useAtomInstance(itemAtom, [
  initialValue
]).exports;
const value = useAtomValue(itemAtom, [initialValue]);
```

After:

```ts
const itemInstance = useAtomInstance(itemAtom, [initialValue]);
const value = useAtomValue(itemInstance); // pass the instance
const { decrement, increment } = itemInstance.exports;
```

This is probably how I would have written it. However, in this particular case, there's a shorthand you can use with [`useAtomState()`](https://omnistac.github.io/zedux/docs/api/hooks/useAtomState):

```ts
const [value, { decrement, increment }] = useAtomState(itemAtom, [
  initialValue,
]);
```

**Reply by devnomic** [reply]
Thank you for your detailed reply again! 
I tried the `useSetAtom` it works very well!  
Just to confirm, the `OPERATION` variable value is `useSetAtom` right?

Yes, if possible please add  `useSetAtom` as utils. People who use jotai will understand immediately as this built in there as counterpart of `useAtomValue` (to set state without rerender).
Also very nice shorthand to set simple state IMO. especially you have more than 1 states to set in one component (less verbose than useAtomInstance).

They can choose to use  `useAtomInstance` setState if they want to go more low level with more versatility.

--- 

````
const [value, { decrement, increment }] = useAtomState(itemAtom, [
  initialValue,
]);
````
Thank you, this is very nice.

I will explore more about zedux!

**Reply by bowheart** [reply]
> Just to confirm, the OPERATION variable value is useSetAtom right?

Right, sorry, should have made it more clear that that's an arbitrary value supplied by you. It's completely optional, but yes, I'd set it to `'useSetAtom'`. It's just to help with debugging - that operation shows up in the atom graph when inspecting it via [`ecosystem.viewGraph`](https://omnistac.github.io/zedux/docs/api/classes/Ecosystem#viewgraph) and in the dev tools (which we haven't released yet). It's like an explanation of why a graph edge was created.

Cool, thanks for the questions. I'll definitely leave this discussion open for a while until I put some of this info into more official documentation/content somewhere.
